### PR TITLE
Fix secret key decryption

### PR DESCRIPTION
**Pull request by sug0**
Fixes decryption of secret keys

### CODE REVIEW

**Comment by badboy** [comment]
Can you expand on why you think this is broken? (I haven't looked at the code in a while so I might be missing context)

File: libsignify/src/encoding.rs [hunk 1]
```diff
@@ -141,7 +141,9 @@ impl Codeable for PrivateKey {
         buf.read_exact(&mut complete_key[..])?;
 
         // sanity check the keypair wasn't corrupted, mostly if it was unencrypted.
-        PrivateKey::verify_keypair(&complete_key)?;
+        if kdf_rounds == 0 {
+            PrivateKey::verify_keypair(&complete_key)?;
+        }
```
Comment (sug0) [inline libsignify/src/encoding.rs:146]: so I think before this change, this check would have never passed for encrypted keys. the reason is the public key derived from the encrypted secret key would never match the decrypted public key. I was hitting this issue when decoding encrypted secret keys.

File: libsignify/src/key.rs [hunk 2]
```diff
@@ -119,7 +119,7 @@ impl PrivateKey {
         } = &derivation_info
         {
             let kdf_rounds = *kdf_rounds;
-            Self::inner_kdf_mix(&mut complete_key.0[..32], kdf_rounds, &salt, passphrase)?;
+            Self::inner_kdf_mix(&mut complete_key.0[..], kdf_rounds, &salt, passphrase)?;
```
Comment (sug0) [inline libsignify/src/key.rs:122]: for this change and...

File: libsignify/src/key.rs [hunk 3]
```diff
@@ -146,7 +146,7 @@ impl PrivateKey {
         let mut encrypted_key = self.complete_key.clone(); // Cheap :)
 
         match Self::inner_kdf_mix(
-            &mut encrypted_key[..32],
+            &mut encrypted_key[..],
```
Comment (sug0) [inline libsignify/src/key.rs:149]: ...this change, to be honest I'm still not 100% sold on the reason why it fixes the issues I was having during decryption. what it essentially fixes is XOR-ing the complete key with the derived key from kdf vs. only XOR-ing the secret key bytes, I believe. before I ran into `Error:WrongKey` I think.

**Review by sug0 — COMMENTED** [review]
Hey @badboy, sure!

File: libsignify/src/encoding.rs (hunk 1 above)
Comment (BlackHoleFox) [reply libsignify/src/encoding.rs:146]: Yes this seems like a bug in hindsight of writing it. I even seem to have ignored the second part of my own comment :D It makes no sense to try validating a ciphertext blob.

File: libsignify/src/key.rs (hunk 3 above)
Comment (BlackHoleFox) [reply libsignify/src/key.rs:149]: I'm wondering if I've missed a detail or if there is something amiss elsewhere if that's the case. From my knowledge it should just be the [secret key](https://github.com/aperezdc/signify/blob/0223172beb99793c4bcc4ea6f6fc4d14dc53036a/signify.c#L309) part that gets [passed into XOR](https://github.com/aperezdc/signify/blob/0223172beb99793c4bcc4ea6f6fc4d14dc53036a/signify.c#L327).

~~What error are you getting back, `WrongKey` or `BadPassword`?~~ whoops can't read today.

File: libsignify/src/encoding.rs (hunk 1 above)
Comment (BlackHoleFox) [reply libsignify/src/encoding.rs:146]: Do you mind adding a regression test of `PrivateKey::from_bytes` that checks both cases of this branch? (doesn't need to test a failed decryption, just that it is only ran when `kdf_rounds` is set appropriately)

Comment (BlackHoleFox) [reply libsignify/src/encoding.rs:146]: PRed this as its own fix in https://github.com/badboy/signify-rs/pull/61

Comment (sug0) [reply libsignify/src/encoding.rs:146]: thanks, I totally forgot about this, as life got in the way :sweat_smile: my bad!

**Pull request closed without merging**